File: demo-test/src/test/java/io/fries/demo/test/cucumber/world/QueryParameter.java
package io.fries.demo.test.cucumber.world;

import java.net.URLEncoder;

import static java.nio.charset.StandardCharsets.UTF_8;
import static java.util.Objects.requireNonNull;

public record QueryParameter(
        String name,
        String value
) {

    public QueryParameter(final String name, final String value) {
        this.name = requireNonNull(name, "name cannot be null");
        this.value = URLEncoder.encode(
                requireNonNull(value, "value cannot be null"),
                UTF_8
        );
    }
}

File: demo-test/src/test/kotlin/io/fries/demo/test/cucumber/world/request/QueryParameter.kt
package io.fries.demo.test.cucumber.world.request

import java.net.URLEncoder
import java.nio.charset.StandardCharsets.UTF_8

data class QueryParameter(
    val name: String,
    private val rawValue: String
) {
    val encodedValue: String = URLEncoder.encode(rawValue, UTF_8)

    init {
        require(name.isNotBlank()) { "name cannot be blank" }
        require(encodedValue.isNotBlank()) { "value cannot be blank" }
    }
}